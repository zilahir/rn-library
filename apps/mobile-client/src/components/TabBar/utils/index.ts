export { default as icons } from "./icons";

export { onPress } from "./navigations";
