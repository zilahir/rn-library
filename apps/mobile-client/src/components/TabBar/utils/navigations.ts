import { BottomTabBarProps } from "@react-navigation/bottom-tabs";

interface INavigationProps {
  navigation: BottomTabBarProps["navigation"];
  targetKey: string;
}

/**
 *
 * @param root0
 * @param root0.navigation
 * @param root0.target
 * @param root0.isFocused
 * @param root0.targetKey
 * @param root0.targetName
 */
export function onPress({
  navigation,
  targetKey,
  isFocused,
  targetName,
}: INavigationProps & { isFocused: boolean; targetName: string }): void {
  const event = navigation.emit({
    type: "tabPress",
    target: targetKey,
    canPreventDefault: true,
  });

  if (!isFocused && !event.defaultPrevented) {
    navigation.navigate(targetName);
  }
}

/*
export function onLongPress({ navigation, targetKey }: INavigationProps): void {
  navigation.emit({
    type: "tabLongPress",
    target: targetKey,
  });
}
*/
