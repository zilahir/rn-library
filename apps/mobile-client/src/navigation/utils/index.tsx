import { ComponentProps } from "react";
import { BottomTabParamList } from "@app/navigation/types";
import HomeScreen from "@app/screens/HomeScreen";
import ProfileScreen from "@app/screens/ProfileScreen";
import TrendingScreen from "@app/screens/TrendingScreen";
import LibraryScreen from "@app/screens/LibraryScreen";

type ScreenName = keyof BottomTabParamList;

interface BottomTabScreen {
  label?: ScreenName;
  name: ScreenName;
  tabComponent: ComponentProps<any>;
  hideLabel?: boolean;
  isFloating?: boolean;
}

interface IBottomTabScreens {
  screens: BottomTabScreen[];
  getBottomTabBarTabs: () => BottomTabScreen[];
  getNumberOfTabBarTabs: () => number;
}

export const bottomTabScreens: IBottomTabScreens = {
  screens: [
    {
      label: "Home",
      name: "Home",
      tabComponent: HomeScreen,
    },

    {
      label: "Profile",
      name: "Profile",
      tabComponent: ProfileScreen,
    },
    {
      label: "Trending",
      name: "Trending",
      tabComponent: TrendingScreen,
    },
    {
      label: "Library",
      name: "Library",
      tabComponent: LibraryScreen,
    },
  ],
  getBottomTabBarTabs: (): BottomTabScreen[] =>
    bottomTabScreens.screens.map((screen) => ({
      ...screen,
      hideLabel: !!screen.hideLabel,
      isFloating: !!screen.isFloating,
    })),

  getNumberOfTabBarTabs: (): number => bottomTabScreens.screens.length,
};
