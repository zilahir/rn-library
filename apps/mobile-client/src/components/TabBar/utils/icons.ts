import { Feather } from "@expo/vector-icons";
import { ComponentProps } from "react";
import { BottomTabParamList } from "@app/navigation/types";

const icons: Record<
  keyof BottomTabParamList,
  ComponentProps<typeof Feather>["name"]
> = {
  Home: "home",
  Profile: "user",
  Trending: "trending-up",
  Library: "book-open",
  Borrow: "book",
};

export default icons;
