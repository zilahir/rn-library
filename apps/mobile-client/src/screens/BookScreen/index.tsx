import { ReactElement } from "react";
import { View, StyleSheet, Image, ImageStyle, Text } from "react-native";
import { useTheme } from "@shopify/restyle";
import { useTranslation } from "react-i18next";
import { FontAwesome, Feather } from "@expo/vector-icons";
import withSafeAreaView from "@app/components/WithSafeAreaView";
import { RootTabScreenProps } from "@app/navigation/types";
import { Theme } from "@app/theme";
import useBookData from "@app/hooks/useBookData";
import ThemeText from "@app/components/common/Text";
import Button from "@app/components/common/Button";
import Badge from "@app/components/common/Badge";

const styles = ({
  spacing,
  colors,
  borderRadii,
}: {
  spacing: Theme["spacing"];
  colors: Theme["colors"];
  borderRadii: Theme["borderRadii"];
}): ReturnType<typeof StyleSheet.create> =>
  StyleSheet.create({
    bookOuterContainer: {
      flex: 1,
    },
    topContainer: {
      justifyContent: "center",
      alignItems: "center",
      backgroundColor: colors.gray100,
      padding: spacing.m,
    },
    bottomContainer: {
      flex: 1,
      justifyContent: "space-evenly",
      alignItems: "center",
    },
    imageStyle: {
      width: 160,
      aspectRatio: 1 / 1.5,
      borderRadius: borderRadii.m,
    },
    imageContainer: {
      shadowColor: "#000",
      shadowOffset: {
        width: 0,
        height: 5,
      },
      shadowOpacity: 0.36,
      shadowRadius: 6.68,
      elevation: 11,
    },
    detailsContainer: {},
    font: {
      textAlign: "center",
    },
    title: {
      fontSize: 20,
      fontWeight: "bold",
    },
    overViewText: {
      fontSize: 20,
      fontWeight: "bold",
      marginBottom: spacing.s,
    },
    author: {
      fontSize: 15,
    },
    metaContainer: {
      flexDirection: "row",
      alignItems: "center",
      marginHorizontal: spacing.m,
      marginVertical: spacing.m,
      paddingHorizontal: spacing.m,
      paddingVertical: spacing.s,
      justifyContent: "center",
      backgroundColor: colors.gray100,
      borderRadius: borderRadii.m,
      borderWidth: 2,
      borderColor: colors.orange100,
    },
    row: {
      flexDirection: "row",
      justifyContent: "center",
      alignItems: "center",
      gap: spacing.s,
    },
    meta: {
      justifyContent: "center",
      alignItems: "center",
      alignContent: "stretch",
      alignSelf: "stretch",
      flex: 1,
      gap: spacing.s,
    },
    rentButtonContainer: {
      alignSelf: "stretch",
      paddingHorizontal: spacing.m,
    },
    overViewContainer: {
      paddingHorizontal: spacing.m,
    },
    avaliableIcon: {
      marginRight: spacing.s,
    },
  });

/**
 *
 * @param {object} root0 - props
 * @param {object} root0.navigation - navigation object passed from react-navigation
 * @param {object} root0.route - route object passed from react-navigation
 * @returns {ReactElement} - react component
 * @description - renders a book screen
 */
function Book({ route, navigation }: RootTabScreenProps<"Book">): ReactElement {
  const { numOfPages, rate, language, overview } = useBookData();
  const { t } = useTranslation();
  const { spacing, colors, borderRadii } = useTheme<Theme>();
  const style = styles({
    spacing,
    colors,
    borderRadii,
  });

  const borrows = route.params.book.borrows ?? [];

  const isAvaliable = borrows.some(({ isBorrowed }) => !isBorrowed);

  return (
    <View style={[style.bookOuterContainer]}>
      <View style={[style.topContainer]}>
        <View style={[style.imageContainer]}>
          <Image
            style={[style.imageStyle as ImageStyle]}
            source={{ uri: route.params.book.cover }}
          />
        </View>
      </View>
      <View style={[style.bottomContainer]}>
        <View style={[style.detailsContainer]}>
          <Text style={[style.font, style.title]}>
            {route.params.book.title}
          </Text>
          <Text style={[style.font, style.author]}>
            {route.params.book.author}
          </Text>
        </View>
        <View style={style.metaContainer}>
          <View style={style.meta}>
            <Text style={[style.metaFont]}>{t("bookMeta.rate")}</Text>
            <View style={[style.row]}>
              <FontAwesome name="star" size={24} color={colors.orange} />
              <Text style={[style.metaFont]}>{rate}</Text>
            </View>
          </View>
          <View style={style.meta}>
            <Text style={[style.metaFont]}>{t("bookMeta.numOfPages")}</Text>
            <Text>{numOfPages}</Text>
          </View>
          <View style={style.meta}>
            <Text style={[style.metaFont]}>{t("bookMeta.language")}</Text>
            <Text>{language}</Text>
          </View>
        </View>
        <View>
          <Badge
            icon={
              <Feather
                style={[style.avaliableIcon]}
                name={isAvaliable ? "x" : "check"}
                color={isAvaliable ? colors.red200 : colors.green100}
              />
            }
            type={isAvaliable ? "errorBadge" : "successBadge"}
            text={t(`bookMeta.isAvaliable.${!isAvaliable}`)}
          />
        </View>
        <View>
          <View style={[style.overViewContainer]}>
            <Text style={[style.overViewText]}>{t("bookMeta.overview")}</Text>
            <ThemeText variant="bookOverview">{overview}</ThemeText>
          </View>
        </View>
        {isAvaliable && (
          <View style={style.rentButtonContainer}>
            <Button
              buttonType="primary"
              label="Rent"
              onPress={(): void =>
                navigation.navigate("BorrowBook", {
                  book: route.params.book,
                })
              }
            />
          </View>
        )}
      </View>
    </View>
  );
}

export default withSafeAreaView(Book);
