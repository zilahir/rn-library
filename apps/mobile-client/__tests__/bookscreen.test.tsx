/* eslint-disable react/jsx-props-no-spreading */
import { fireEvent } from "@testing-library/react-native";
import BookScreen from "@app/screens/BookScreen";

import { render } from "./utils/wrapper";

const props = {
  navigation: {
    navigate: jest.fn(),
  },
  route: {
    params: {
      book: {
        title: "Hello World",
        _id: "5f5f5f5f5f5f5f5f5f5f5f5f",
        isbn: "978-1-56619-909-4",
        author: "John Doe",
        description: "This is a book",
        borrows: [
          {
            book: {
              title: "Hello World",
              _id: "5f5f5f5f5f5f5f5f5f5f5f5f",
              isbn: "978-1-56619-909-4",
              author: "John Doe",
              description: "This is a book",
            },
            isbn: "978-1-56619-909-4",
            isBorrowed: false,
            user: {
              userId: "1",
              _id: "5f5f5f5f5f5f5f5f5f5f5f5f",
            },
          },
        ],
      },
    },
  },
} as any;

describe("BorrowBookScreen", () => {
  it("renders correctly", () => {
    const { toJSON } = render(<BookScreen {...props} />);

    // Use snapshot testing to capture the component's output
    expect(toJSON()).toMatchSnapshot();
  });

  it("camera is defined", () => {
    const { queryByTestId } = render(<BookScreen {...props} />);
    const screen = queryByTestId("borrow-outer-container");

    expect(screen).toBeDefined();
  });

  it("navigation is defined", () => {
    const { getByTestId } = render(<BookScreen {...props} />);

    fireEvent.press(getByTestId("Rent"));

    expect(props.navigation.navigate).toHaveBeenCalled();
  });
});
