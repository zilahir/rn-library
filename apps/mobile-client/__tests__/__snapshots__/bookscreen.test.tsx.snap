// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BorrowBookScreen renders correctly 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#fff",
        "borderBottomLeftRadius": 25,
        "borderBottomRightRadius": 25,
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "rgba(247, 247, 247, 1)",
            "justifyContent": "center",
            "padding": 16,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "elevation": 11,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 5,
                "width": 0,
              },
              "shadowOpacity": 0.36,
              "shadowRadius": 6.68,
            },
          ]
        }
      >
        <Image
          source={
            {
              "uri": undefined,
            }
          }
          style={
            [
              {
                "aspectRatio": 0.6666666666666666,
                "borderRadius": 10,
                "width": 160,
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        [
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "space-evenly",
          },
        ]
      }
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "fontSize": 20,
                "fontWeight": "bold",
              },
            ]
          }
        >
          Hello World
        </Text>
        <Text
          style={
            [
              {
                "textAlign": "center",
              },
              {
                "fontSize": 15,
              },
            ]
          }
        >
          John Doe
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(247, 247, 247, 1)",
            "borderColor": "rgba(255, 153, 0, 0.05)",
            "borderRadius": 10,
            "borderWidth": 2,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginHorizontal": 16,
            "marginVertical": 16,
            "paddingHorizontal": 16,
            "paddingVertical": 8,
          }
        }
      >
        <View
          style={
            {
              "alignContent": "stretch",
              "alignItems": "center",
              "alignSelf": "stretch",
              "flex": 1,
              "gap": 8,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                undefined,
              ]
            }
          >
            bookMeta.rate
          </Text>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 8,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text />
            <Text
              style={
                [
                  undefined,
                ]
              }
            >
              4.0
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "alignContent": "stretch",
              "alignItems": "center",
              "alignSelf": "stretch",
              "flex": 1,
              "gap": 8,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                undefined,
              ]
            }
          >
            bookMeta.numOfPages
          </Text>
          <Text>
<<<<<<< HEAD
            340
=======
            821
>>>>>>> a27136c (bookscreen tests)
          </Text>
        </View>
        <View
          style={
            {
              "alignContent": "stretch",
              "alignItems": "center",
              "alignSelf": "stretch",
              "flex": 1,
              "gap": 8,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                undefined,
              ]
            }
          >
            bookMeta.language
          </Text>
          <Text>
            ðŸ‡«ðŸ‡®
          </Text>
        </View>
      </View>
      <View>
        <View
          style={
            [
              {
                "alignItems": "center",
                "backgroundColor": "#ffcac7",
                "borderRadius": 25,
                "flexDirection": "row",
                "paddingHorizontal": 16,
                "paddingVertical": 8,
              },
            ]
          }
        >
          <Text />
          <Text
            style={
              [
                {
                  "color": "#ff372e",
                  "fontSize": 14,
                },
              ]
            }
          >
            bookMeta.isAvaliable.false
          </Text>
        </View>
      </View>
      <View>
        <View
          style={
            [
              {
                "paddingHorizontal": 16,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 20,
                  "fontWeight": "bold",
                  "marginBottom": 8,
                },
              ]
            }
          >
            bookMeta.overview
          </Text>
          <Text
            style={
              [
                {
                  "lineHeight": 20,
                },
              ]
            }
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "alignSelf": "stretch",
            "paddingHorizontal": 16,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#FF9900",
              "borderRadius": 25,
              "height": 50,
              "justifyContent": "center",
              "opacity": 1,
              "textAlign": "center",
            }
          }
          testID="Rent"
        >
          <Text
            style={
              [
                {
                  "color": "#fff",
                  "fontSize": 16,
                  "fontWeight": 500,
                },
              ]
            }
          >
            Rent
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
